### PR TITLE
Support using same browser across AsyncChatGPTs

### PR DESCRIPTION
**Pull request by SuperKenVery**
Note: sync ChatGPT doesn't support this

### CODE REVIEW

File: chatgpt_wrapper/chatgpt.py [hunk 1]
```diff
@@ -488,6 +392,17 @@ def new_conversation(self):
 class ChatGPT:
 
     def __init__(self, headless: bool = True, browser="firefox", model="default", timeout=60, debug_log=None, proxy: Optional[ProxySettings] = None):
+        """Create ChatGPT
+
+        Args:
+            headless (bool, optional): _description_. Defaults to True.
+            browser (str, optional): The firefox to use. Only support creating new browser. To use existing browser, use `AsyncChatGPT` instead. Defaults to "firefox".
```
Comment (thehunmonkgroup) [inline chatgpt_wrapper/chatgpt.py:399]: The firefox to use? Can't a user use a different browser?

**Review by thehunmonkgroup — CHANGES_REQUESTED** [review]
Can you please provide a written description of what this PR is accomplishing? I did review the code, but a high level description would be helpful to make sure I'm understanding this change correctly.

Also, I think it would be helpful to provide some example code making use of this enhancement.

Almost there!

File: chatgpt_wrapper/chatgpt.py (hunk 1 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/chatgpt.py:399]: Ahh this was a typo 🤣 

But you reminded me... A bug is here too

**Comment by SuperKenVery** [comment]
> Can you please provide a written description of what this PR is accomplishing

Sure! 

Currently when you creeate an `AsyncChatGPT` instance you create a new browser along with it. This PR allows you to reuse an existing browser, page and session (but you still have multiple conversations because conversation id is still unique to every `AsyncChatGPT` instance). 

In this way we can save some memory and speed up creation of `AsyncChatGPT` instances.

**Comment by SuperKenVery** [comment]
Please have a look. Now it's not very convinient for me to test because of some network issues 😢 but I think it should work.

**Comment by thehunmonkgroup** [comment]
Apologies for not mentioning this earlier, but can you remove the logging abstraction you added? That seems out of scope for this PR, plus there are some other changes in the works that will probably obviate that approach.

In general, I think it's best to keep PRs very focused on one task at a time. I know it's tempting to start refactoring while you're already in there (I do it too), but as a maintainer reviewing these PRs, it makes it harder to review and reason about, and increases the chances it will conflict with somebody else's work.

Perhaps instead of that abstraction, we need a base class for the `ChatGPT` classes?

File: chatgpt_wrapper/browser.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        if AsyncBrowser.play is None:
+            AsyncBrowser.play=await async_playwright().start()
+
+        if not hasattr(self.play,browser):
+            self.log.error("Browser %s is invalid, falling back on firefox. ",browser)
+            browser='firefox'
+        self.browser:BrowserType=getattr(self.play,browser)
+
+        try:
+            self.browser_context=await self.browser.lauch_persistent_context(
```
Comment (thehunmonkgroup) [inline chatgpt_wrapper/browser.py:33]: Spelling error.

**Review by thehunmonkgroup — CHANGES_REQUESTED** [review]
After fixing the spelling error, when trying to test with this code:

```python
#!/usr/bin/env python

import asyncio
from chatgpt_wrapper import AsyncChatGPT

async def main():
    gpt = AsyncChatGPT()
    bot1 = await gpt.create(False)
    #bot2 = await gpt.create(browser=bot1.browser)

    response1 = await bot1.ask("Say, bot1!")
    #response2 = await bot2.ask("Say, bot2!")
    print(response1)
    #print(response2)

asyncio.run(main())
```

I received this error:

```
  File "/home/hunmonk/git/chatgpt-wrapper/tests/test_async_chatgpt.py", line 16, in <module>
    asyncio.run(main())
  File "/home/hunmonk/.pyenv/versions/3.10.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/home/hunmonk/.pyenv/versions/3.10.10/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/home/hunmonk/git/chatgpt-wrapper/tests/test_async_chatgpt.py", line 8, in main
    bot1 = await gpt.create(False)
  File "/home/hunmonk/git/chatgpt-wrapper/chatgpt_wrapper/chatgpt.py", line 53, in create
    self.browser=browser if browser else await AsyncBrowser().create(headless=headless,proxy=proxy,debug_log=debug_log)
  File "/home/hunmonk/git/chatgpt-wrapper/chatgpt_wrapper/browser.py", line 48, in create
    await self.goto_chat()
  File "/home/hunmonk/git/chatgpt-wrapper/chatgpt_wrapper/browser.py", line 58, in goto_chat
    raise err
  File "/home/hunmonk/git/chatgpt-wrapper/chatgpt_wrapper/browser.py", line 54, in goto_chat
    await self.page.goto("https://chat.openai.com",timeout=timeout)
  File "/home/hunmonk/.pyenv/versions/3.10.10/lib/python3.10/site-packages/playwright/async_api/_generated.py", line 9162, in goto
    await self._impl_obj.goto(
  File "/home/hunmonk/.pyenv/versions/3.10.10/lib/python3.10/site-packages/playwright/_impl/_page.py", line 494, in goto
    return await self._main_frame.goto(**locals_to_params(locals()))
  File "/home/hunmonk/.pyenv/versions/3.10.10/lib/python3.10/site-packages/playwright/_impl/_frame.py", line 147, in goto
    await self._channel.send("goto", locals_to_params(locals()))
  File "/home/hunmonk/.pyenv/versions/3.10.10/lib/python3.10/site-packages/playwright/_impl/_connection.py", line 44, in send
    return await self._connection.wrap_api_call(
  File "/home/hunmonk/.pyenv/versions/3.10.10/lib/python3.10/site-packages/playwright/_impl/_connection.py", line 419, in wrap_api_call
    return await cb()
  File "/home/hunmonk/.pyenv/versions/3.10.10/lib/python3.10/site-packages/playwright/_impl/_connection.py", line 79, in inner_send
    result = next(iter(done)).result()
playwright._impl._api_types.TimeoutError: Timeout 15ms exceeded.
=========================== logs ===========================
navigating to "https://chat.openai.com/", waiting until "load"
```

I think I'm going to let you get this working in your environment before I go any further!

**Comment by SuperKenVery** [comment]
> That seems out of scope for this PR

I truly understand that a PR should be minimal. However if I don't add this abstraction I would have to copy a lot of things from `AsyncChatGPT` to `Browser`, which I think is not elegant. What do you think? 

> I think I'm going to let you get this working in your environment before I go any further!

Sure! Will try to do that!

File: chatgpt_wrapper/browser.py (hunk 2 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/browser.py:33]: Sorry about my poor English but I can't see any spelling error here...
And this part code runs on my setup (after fixing bugs somewhere else).

**Comment by SuperKenVery** [comment]
Could you please have a look? 
If you think I should break this into 2 PRs just tell me then~

**Comment by thehunmonkgroup** [comment]
See https://github.com/mmabrouk/chatgpt-wrapper/commit/31dab8dabe1822cea3bde6686347fc7434a919d7

Guessing with the last few commits I did today you're going to need to refactor anyways.

**Pull request closed without merging**